src/libraries/System.Private.CoreLib/src/System/Text/TranscodingStream.cs

```diff
@@ -0,0 +1,603 @@
+                byteCount = _innerEncoder.GetBytes(chars, 0, charCount, bytes, 0, flush: true);
+            }
+
+            return new ArraySegment<byte>(bytes, 0, byteCount);
+        }
+
+        public override void Flush()
+        {
+            // Don't pass flush: true to our inner decoder + encoder here, since it could cause data
+            // corruption if a flush occurs mid-stream. Wait until the stream is being closed.
+
+            ThrowIfDisposed();
+            _innerStream.Flush();
+        }
+
+        public override Task FlushAsync(CancellationToken cancellationToken)
+        {
+            // Don't pass flush: true to our inner decoder + encoder here, since it could cause data
+            // corruption if a flush occurs mid-stream. Wait until the stream is being closed.
+
+            ThrowIfDisposed();
+            return _innerStream.FlushAsync(cancellationToken);
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (buffer is null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            return Read(new Span<byte>(buffer, offset, count));
+        }
+
+        public override int Read(Span<byte> buffer)
+        {
+            EnsurePreReadConditions();
+
+            // If there's no data in our pending read buffer, we'll need to populate it from
+            // the inner stream. We read the inner stream's bytes, decode that to chars using
+            // the 'inner' encoding, then re-encode those chars under the 'this' encoding.
+            // We've already calculated the worst-case expansions for the intermediate buffers,
+            // so we use GetChars / GetBytes instead of Convert to simplify the below code
+            // and to ensure an exception is thrown if the Encoding reported an incorrect
+            // worst-case expansion.
+
+            if (_readBufferCount == 0)
+            {
+                byte[] rentedBytes = ArrayPool<byte>.Shared.Rent(DefaultReadByteBufferSize);
+                char[] rentedChars = ArrayPool<char>.Shared.Rent(_readCharBufferMaxSize);
+
+                try
+                {
+                    // Beware: Use our constant value instead of 'rentedBytes.Length' for the count
+                    // parameter below. The reason for this is that the array pool could've returned
+                    // a larger-than-expected array, but our worst-case expansion calculations
+                    // performed earlier didn't take that into account.
+
+                    int innerBytesReadJustNow = _innerStream.Read(rentedBytes, 0, DefaultReadByteBufferSize);
+                    bool isEofReached = (innerBytesReadJustNow == 0);
+
+                    // convert bytes [inner] -> chars, then convert chars -> bytes [this]
+
+                    int charsDecodedJustNow = _innerDecoder.GetChars(rentedBytes, 0, innerBytesReadJustNow, rentedChars, 0, flush: isEofReached);
```

Comment:
No. [Check out the implementation](https://github.com/dotnet/runtime/blob/a5414a2bdf5a3e45e3800530bfd35de128971ac1/src/libraries/System.Private.CoreLib/src/System/Text/Decoder.cs) of the `Decoder` class. The only abstract method is the `byte[]`-accepting method. Both the span overloads and the pointer overloads are written in terms of the array overload, but they need to allocate temporary arrays in order to work. Since the array overload is the only one that the subclassed type is guaranteed to have implemented efficiently, and since we already have arrays, we may as well call the array-based overload instead of the possibly less efficient span-based overload.

Reply:
I see, in that case the same could be said from the `Convert` methods used for Write/WriteAsync, right?
https://github.com/dotnet/runtime/pull/35145/files/9c6f7d69378b1fe4121bfb67908fbb7ca0ff3407#diff-20de3ee0422a4fcbf73bdc5463bbe123R563

You are funneling array-based into span-based. Wouldn't be more efficient the other way around?

Reply:
That's a good observation! In the `Write` case, we're not guaranteed already to have an array. Write buffers can be provided as either a `byte[]` or a `ROS<byte>`. We could split out those code paths to special-case `byte[]` vs. `ROS<byte>`, but now we have two essentially duplicate implementations of the same method, just to handle a scenario where somebody has written a custom `Encoder` or `Decoder`. That doesn't seem worth the engineering tradeoff.

(I can change the call to `_innerEncoder.Convert` relatively simply without duplicating code, but not the call to `_thisDecoder.Convert`.)

In the `Read` case, since we're already guaranteed to have backing arrays, we may as well call the array-accepting overloads. There's no code duplication required.

Reply:
Necessary context for my above comments: all in-box implementations of the `Encoding` class provides `Encoder` or `Decoder` instances which override both the array-based and the span-based conversion routines. So the extra temporary allocation only applies when somebody has written a completely custom `Encoding` class, which almost never happens. That's why I'm not too worried about this.

That said, on the off-chance somebody _did_ write such a class, and if we can provide that scenario optimal performance by changing just a single line of code on our end, then sure, we may as well go for it. :)
But it's uncommon enough that I wouldn't really be willing to change more than one or two lines of code for it.

Reply:
Since code duplication is the only way to avoid calling the slightly worse base implementation, which also seems unlikely to be used because it gets overridden in all of our in-box Encoders, I lean towards to keep the code as it is right now.

Thanks for elaborating.